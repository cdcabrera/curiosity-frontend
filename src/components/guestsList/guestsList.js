import React from 'react';
import PropTypes from 'prop-types';
import { TableVariant } from '@patternfly/react-table';
import { helpers } from '../../common';
import { apiQueries, connect, reduxActions, reduxSelectors } from '../../redux';
import { Loader } from '../loader/loader';
import { inventoryListHelpers } from '../inventoryList/inventoryListHelpers';
import { RHSM_API_QUERY_TYPES } from '../../types/rhsmApiTypes';
import { Table } from '../table/table';

/**
 * A system inventory guests component.
 *
 * @augments React.Component
 * @fires onUpdateGuestsData
 * @fires onScroll
 */
class GuestsList extends React.Component {
  state = { currentPage: 0, limit: 10, previousData: [] };

  componentDidMount() {
    this.onUpdateGuestsData();
  }

  componentDidUpdate(prevProps, prevState) {
    const { currentPage } = this.state;

    if (currentPage !== prevState.currentPage) {
      this.onUpdateGuestsData();
    }
  }

  /**
   * Call the RHSM APIs, apply filters.
   *
   * @event onUpdateGuestsData
   */
  onUpdateGuestsData = async () => {
    const { currentPage, limit } = this.state;
    const { getHostsInventoryGuests, query, id } = this.props;

    if (id) {
      const updatedQuery = {
        ...query,
        [RHSM_API_QUERY_TYPES.LIMIT]: limit,
        [RHSM_API_QUERY_TYPES.OFFSET]: currentPage * limit || 0
      };

      const { guestsQuery } = apiQueries.parseRhsmQuery(updatedQuery);
      await getHostsInventoryGuests(id, guestsQuery);
    }
  };

  /**
   * Update page state.
   *
   * @event onScroll
   * @param {object} event
   */
  onScroll = event => {
    const { target } = event;
    const { currentPage, limit, previousData } = this.state;
    const { numberOfEntries, pending, listData } = this.props;

    // const distanceBottom = target.scrollHeight - (target.scrollTop + target.clientHeight);
    const bottom = target.scrollHeight - target.scrollTop === target.clientHeight;

    if (numberOfEntries > (currentPage + 1) * limit && bottom && !pending) {
      const newPage = currentPage + 1;
      const updatedData = [...previousData, ...(listData || [])];

      this.setState({
        previousData: updatedData,
        currentPage: newPage
      });
    }
  };

  renderLoader() {
    const { currentPage } = this.state;
    const { pending } = this.props;

    if (currentPage > 0 && pending) {
      return (
        <div className="curiosity-pagination-scroll-loader__spinner">
          <Loader />
        </div>
      );
    }

    return null;
  }

  /**
   * Render a guests table.
   *
   * @returns {Node}
   */
  renderTable() {
    const { previousData } = this.state;
    const { filterGuestsData, listData } = this.props;
    let updatedColumnHeaders = [];

    const updatedRows = [...previousData, ...(listData || [])].map(({ ...cellData }) => {
      const { columnHeaders, cells } = inventoryListHelpers.parseRowCellsListData({
        filters: filterGuestsData,
        cellData
      });

      updatedColumnHeaders = columnHeaders;

      return {
        cells
      };
    });

    return (
      <div className="curiosity-pagination-scroll" style={{ height: `200px` }}>
        <div className="curiosity-pagination-scroll-list" onScroll={this.onScroll}>
          {this.renderLoader()}
          {(updatedRows.length && (
            <Table
              borders={false}
              variant={TableVariant.compact}
              className="curiosity-inventory-list"
              columnHeaders={updatedColumnHeaders}
              rows={updatedRows}
            />
          )) ||
            null}
        </div>
      </div>
    );
  }

  /**
   * Render a guest list table.
   *
   * @returns {Node}
   */
  render() {
    const { currentPage } = this.state;
    const { error, filterGuestsData, listData, pending, perPageDefault } = this.props;

    return (
      <div className={`fadein ${(error && 'blur') || ''}`}>
        {pending && currentPage === 0 && (
          <Loader
            variant="table"
            tableProps={{
              borders: false,
              colCount: filterGuestsData?.length || (listData?.[0] && Object.keys(listData[0]).length) || 1,
              rowCount: perPageDefault,
              variant: TableVariant.compact
            }}
          />
        )}
        {((!pending && currentPage === 0) || currentPage > 0) && this.renderTable()}
      </div>
    );
  }
}

/**
 * Prop types.
 *
 * @type {{numberOfEntries: number, listData: Array, getHostsInventoryGuests: Function,
 *     filterGuestsData: object, pending: boolean, query: object, perPageDefault: number,
 *     id: string, error: boolean}}
 */
GuestsList.propTypes = {
  error: PropTypes.bool,
  filterGuestsData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      header: PropTypes.oneOfType([
        PropTypes.shape({
          title: PropTypes.string
        }),
        PropTypes.func
      ]),
      cell: PropTypes.oneOfType([
        PropTypes.shape({
          title: PropTypes.string
        }),
        PropTypes.func
      ])
    }).isRequired
  ),
  getHostsInventoryGuests: PropTypes.func,
  numberOfEntries: PropTypes.number,
  listData: PropTypes.array,
  pending: PropTypes.bool,
  perPageDefault: PropTypes.number,
  query: PropTypes.object.isRequired,
  id: PropTypes.string.isRequired
};

/**
 * Default props.
 *
 * @type {{numberOfEntries: number, listData: Array, getHostsInventoryGuests: Function,
 *     filterGuestsData: Array, pending: boolean, perPageDefault: number, error: boolean}}
 */
GuestsList.defaultProps = {
  error: false,
  filterGuestsData: [],
  getHostsInventoryGuests: helpers.noop,
  numberOfEntries: 0,
  listData: [],
  pending: false,
  perPageDefault: 5
};

/**
 * Apply actions to props.
 *
 * @param {Function} dispatch
 * @returns {object}
 */
const mapDispatchToProps = dispatch => ({
  getHostsInventoryGuests: (id, query) => dispatch(reduxActions.rhsm.getHostsInventoryGuests(id, query))
});

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.guestsList.makeGuestsList();

const ConnectedGuestsList = connect(makeMapStateToProps, mapDispatchToProps)(GuestsList);

export { ConnectedGuestsList as default, ConnectedGuestsList, GuestsList };
