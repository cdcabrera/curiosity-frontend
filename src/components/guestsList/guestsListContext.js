import { useUnmount, useShallowCompareEffect } from 'react-use';
import { reduxActions, reduxTypes, storeHooks } from '../../redux';
import { useProductInventoryGuestsConfig, useProductInventoryGuestsQuery } from '../productView/productViewContext';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';

/**
 * Default paging limit for guests inventory.
 *
 * @type {number}
 */
const DEFAULT_LIMIT = 100;

/**
 * Default paging page for guests inventory.
 *
 * @type {number}
 */
const DEFAULT_OFFSET = 0;

/**
 * Guests inventory selector response.
 *
 * @param {string} id
 * @param {object} options
 * @param {Function} options.useSelectorsResponse
 * @returns {{data: (*|{}), pending: (*|boolean), fulfilled, error}}
 */
const useSelectorsGuestsInventory = (
  id,
  { useSelectorsResponse: useAliasSelectorsResponse = storeHooks.reactRedux.useSelectorsResponse } = {}
) => {
  const { error, cancelled, fulfilled, pending, data } = useAliasSelectorsResponse(
    ({ inventory }) => inventory?.hostsGuests?.[id]
  );

  return {
    error,
    fulfilled,
    pending: pending || cancelled || false,
    data: (data?.length === 1 && data[0]) || data || {}
  };
};

/**
 * Combined Redux RHSM Actions, getHostsInventoryGuests, and inventory selector response.
 *
 * @param {string} id
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useProductInventoryGuestsQuery
 * @param {Function} options.useSelectorsGuestsInventory
 * @returns {Function}
 */
const useGetGuestsInventory = (
  id,
  {
    useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
    useProductInventoryGuestsQuery: useAliasProductInventoryGuestsQuery = useProductInventoryGuestsQuery,
    useSelectorsGuestsInventory: useAliasSelectorsGuestsInventory = useSelectorsGuestsInventory
  } = {}
) => {
  const query = useAliasProductInventoryGuestsQuery({ options: { overrideId: id } });
  const dispatch = useAliasDispatch();
  const response = useAliasSelectorsGuestsInventory(id);

  useShallowCompareEffect(() => {
    reduxActions.rhsm.getHostsInventoryGuests(id, query)(dispatch);
  }, [dispatch, id, query]);

  return {
    ...response
  };
};

/**
 * Use paging as onScroll event for guests inventory.
 *
 * @param {string} id
 * @param {Function} successCallback
 * @param {object} options
 * @param {number} options.defaultLimit
 * @param {number} options.defaultOffset
 * @param {Function} options.useDispatch
 * @param {Function} options.useProductInventoryGuestsConfig
 * @param {Function} options.useProductInventoryGuestsQuery
 * @param {Function} options.useSelectorsGuestsInventory
 * @returns {Function}
 */
const useOnScroll = (
  id,
  successCallback,
  {
    defaultLimit = DEFAULT_LIMIT,
    defaultOffset = DEFAULT_OFFSET,
    useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
    useSelectorsGuestsInventory: useAliasSelectorsGuestsInventory = useSelectorsGuestsInventory,
    useProductInventoryGuestsConfig: useAliasProductInventoryGuestsConfig = useProductInventoryGuestsConfig,
    useProductInventoryGuestsQuery: useAliasProductInventoryGuestsQuery = useProductInventoryGuestsQuery
  } = {}
) => {
  const { initialQuery } = useAliasProductInventoryGuestsConfig();
  const dispatch = useAliasDispatch();
  const { pending, data = {} } = useAliasSelectorsGuestsInventory(id);
  const { count: numberOfGuests } = data?.meta || {};

  const query = useAliasProductInventoryGuestsQuery({ options: { overrideId: id } });
  const {
    [RHSM_API_QUERY_SET_TYPES.LIMIT]: limit = initialQuery[RHSM_API_QUERY_SET_TYPES.LIMIT] || defaultLimit,
    [RHSM_API_QUERY_SET_TYPES.OFFSET]: currentPage = initialQuery[RHSM_API_QUERY_SET_TYPES.OFFSET] || defaultOffset
  } = query;

  /**
   * Reset paging in scenarios where inventory is filtered, or guests is collapsed.
   */
  useUnmount(() => {
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RHSM_GUESTS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.OFFSET],
        viewId: id,
        [RHSM_API_QUERY_SET_TYPES.OFFSET]: initialQuery[RHSM_API_QUERY_SET_TYPES.OFFSET] || defaultOffset
      }
    ]);
  });

  /**
   * On scroll, dispatch type.
   *
   * @event onScroll
   * @param {object} event
   * @returns {void}
   */
  return event => {
    const { target } = event;
    const bottom = target.scrollHeight - target.scrollTop === target.clientHeight;

    if (numberOfGuests > (currentPage + 1) * limit && bottom && !pending) {
      if (typeof successCallback === 'function') {
        successCallback(event);
      }

      dispatch([
        {
          type: reduxTypes.query.SET_QUERY_RHSM_GUESTS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.OFFSET],
          viewId: id,
          [RHSM_API_QUERY_SET_TYPES.OFFSET]: currentPage + 1
        },
        {
          type: reduxTypes.query.SET_QUERY_RHSM_GUESTS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.LIMIT],
          viewId: id,
          [RHSM_API_QUERY_SET_TYPES.LIMIT]: limit
        }
      ]);
    }
  };
};

const context = {
  DEFAULT_LIMIT,
  DEFAULT_OFFSET,
  useGetGuestsInventory,
  useOnScroll,
  useSelectorsGuestsInventory
};

export {
  context as default,
  context,
  DEFAULT_LIMIT,
  DEFAULT_OFFSET,
  useGetGuestsInventory,
  useOnScroll,
  useSelectorsGuestsInventory
};
