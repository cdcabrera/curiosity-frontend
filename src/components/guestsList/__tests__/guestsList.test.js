import React from 'react';
import { shallow } from 'enzyme';
import { GuestsList } from '../guestsList';

describe('GuestsList Component', () => {
  it('should render a non-connected component', () => {
    const props = {
      id: 'lorem'
    };

    const component = shallow(<GuestsList {...props} />);
    expect(component).toMatchSnapshot('non-connected');
  });

  it('should handle variations in data', () => {
    const props = {
      id: 'lorem',
      listData: [
        { lorem: 'ipsum', dolor: 'sit' },
        { lorem: 'sit', dolor: 'amet' }
      ]
    };

    const component = shallow(<GuestsList {...props} />);
    expect(component).toMatchSnapshot('variable data');

    component.setProps({
      filterGuestsData: [{ id: 'lorem' }]
    });

    expect(component).toMatchSnapshot('filtered data');
  });
});
