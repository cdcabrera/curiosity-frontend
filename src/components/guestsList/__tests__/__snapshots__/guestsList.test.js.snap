// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuestsList Component should handle variations in data: filtered data 1`] = `
<div
  className="fadein "
>
  <div
    className="curiosity-table-scroll"
    style={
      Object {
        "height": "200px",
      }
    }
  >
    <div
      className="curiosity-table-scroll-list"
      onScroll={[Function]}
    >
      <Table
        ariaLabel={null}
        borders={false}
        className="curiosity-inventory-list"
        columnHeaders={
          Array [
            "t(curiosity-inventory.header, [object Object])",
          ]
        }
        isHeader={true}
        rows={
          Array [
            Object {
              "cells": Array [
                "ipsum",
              ],
            },
            Object {
              "cells": Array [
                "sit",
              ],
            },
          ]
        }
        summary={null}
        t={[Function]}
        variant="compact"
      />
    </div>
  </div>
</div>
`;

exports[`GuestsList Component should handle variations in data: variable data 1`] = `
<div
  className="fadein "
>
  <div
    className="curiosity-table-scroll"
    style={
      Object {
        "height": "200px",
      }
    }
  >
    <div
      className="curiosity-table-scroll-list"
      onScroll={[Function]}
    >
      <Table
        ariaLabel={null}
        borders={false}
        className="curiosity-inventory-list"
        columnHeaders={
          Array [
            "t(curiosity-inventory.header, [object Object])",
            "t(curiosity-inventory.header, [object Object])",
          ]
        }
        isHeader={true}
        rows={
          Array [
            Object {
              "cells": Array [
                "ipsum",
                "sit",
              ],
            },
            Object {
              "cells": Array [
                "sit",
                "amet",
              ],
            },
          ]
        }
        summary={null}
        t={[Function]}
        variant="compact"
      />
    </div>
  </div>
</div>
`;

exports[`GuestsList Component should render a non-connected component: non-connected 1`] = `
<div
  className="fadein "
>
  <div
    className="curiosity-table-scroll"
    style={
      Object {
        "height": "200px",
      }
    }
  >
    <div
      className="curiosity-table-scroll-list"
      onScroll={[Function]}
    />
  </div>
</div>
`;
